src/main/kotlin/org/treeWare/proto3/message/EncodeProto3.kt

```diff
@@ -0,0 +1,214 @@
+    }
+
+    override fun leaveEnumerationMeta(leaderEnumerationMeta1: EntityModel) {
+        bodyWriter.writeln_l("}")
+        bodyWriter.writeln("")
+    }
+
+    override fun visitEnumerationValueMeta(leaderEnumerationValueMeta1: EntityModel): TraversalAction {
+        val enumValue = getMetaName(leaderEnumerationValueMeta1).uppercase()
+        val enumNumber = getMetaNumber(leaderEnumerationValueMeta1)
+        bodyWriter.writeln("$enumValue = $enumNumber")
+        return TraversalAction.CONTINUE
+    }
+
+    override fun leaveEnumerationValueMeta(leaderEnumerationValueMeta1: EntityModel) {
+
+    }
+
+    override fun visitEntityMeta(leaderEntityMeta1: EntityModel): TraversalAction {
+        val entityName = getMetaName(leaderEntityMeta1)
+        bodyWriter.writeln_r("message ${snakeToCamel(entityName)} {")
+        visitMessageOptions(leaderEntityMeta1)
+        return TraversalAction.CONTINUE
+    }
+
+    override fun leaveEntityMeta(leaderEntityMeta1: EntityModel) {
+        bodyWriter.writeln_l("}")
+        bodyWriter.writeln()
+    }
+
+    override fun visitFieldMeta(leaderFieldMeta1: EntityModel): TraversalAction {
+        val fieldIndex = getMetaNumber(leaderFieldMeta1)
+        val name = snakeToLowerCamel(getMetaName(leaderFieldMeta1))
+        val type = getEncodingType(leaderFieldMeta1)
+        var repeat = ""
+
+        if (getMultiplicityMeta(leaderFieldMeta1) == Multiplicity.LIST ||
+            getMultiplicityMeta(leaderFieldMeta1) == Multiplicity.SET
+        ) repeat = "repeated "
+
+        bodyWriter.write("$repeat$type $name = $fieldIndex")
+        visitFieldOptions(leaderFieldMeta1)
+
+        val fieldType = getFieldTypeMeta(leaderFieldMeta1)
+        if (fieldType == FieldType.COMPOSITION || fieldType == FieldType.ENUMERATION) {
+            val resolvedEntity: EntityModel =
+                if (fieldType == FieldType.COMPOSITION) {
+                    checkNotNull(getMetaModelResolved(leaderFieldMeta1)?.compositionMeta)
+                } else {
+                    checkNotNull(getMetaModelResolved(leaderFieldMeta1)?.enumerationMeta)
+                }
+
+            val packageName = getPackageName(resolvedEntity)
+            if (packageName != currentPackage) {
+                val importLine = "import \"$packageName.proto\";"
+                if (!headerWriter.toString().contains(importLine))
+                    headerWriter.writeln("import \"$packageName.proto\";")
+            }
+        }
+        return TraversalAction.CONTINUE
+    }
+
+    override fun leaveFieldMeta(leaderFieldMeta1: EntityModel) {
+    }
+
+    fun visitMessageOptions(leaderFieldMeta: EntityModel) {
+        getProto3MetaModelMap(leaderFieldMeta)?.options?.forEach { option ->
+            bodyWriter.writeln("option $option;")
+        }
+    }
+
+    fun visitFieldOptions(leaderFieldMeta: EntityModel) {
+        var hasWrittenFirstOption = false
+
+        if (isKeyFieldMeta(leaderFieldMeta)) {
```

Comment:
nit: would be cleaner to combine the two lists (keyFieldOptions + options or requiredFieldOptions + options), and then use the `joinToString()` method on that list.